tests/callbacks/test_param_update_callbacks.py

```diff
@@ -1,33 +1,56 @@
 from copy import deepcopy
 
+import pytest
 import torch
 from torch import nn
 
 from pl_bolts.callbacks.byol_updates import BYOLMAWeightUpdate
 
 
-def test_byol_ma_weight_update_callback():
-    a = nn.Linear(100, 10)
-    b = deepcopy(a)
-    a_original = deepcopy(a)
-    b_original = deepcopy(b)
-
-    # make sure a params and b params are the same
-    assert torch.equal(next(iter(a.parameters()))[0], next(iter(b.parameters()))[0])
-
-    # fake weight update
-    opt = torch.optim.SGD(a.parameters(), lr=0.1)
-    y = a(torch.randn(3, 100))
-    loss = y.sum()
-    loss.backward()
-    opt.step()
-    opt.zero_grad()
-
-    # make sure a did in fact update
-    assert not torch.equal(next(iter(a_original.parameters()))[0], next(iter(a.parameters()))[0])
-
-    # do update via callback
-    cb = BYOLMAWeightUpdate(0.8)
-    cb.update_weights(a, b)
-
-    assert not torch.equal(next(iter(b_original.parameters()))[0], next(iter(b.parameters()))[0])
+@pytest.mark.parametrize("initial_tau", [-0.1, 0.0, 0.996, 1.0, 1.1])
+def test_byol_ma_weight_single_update_callback(initial_tau):
+    """Check BYOL exponential moving average weight update rule for a single update."""
+    if not 0.0 <= initial_tau <= 1.0:
+        with pytest.raises(ValueError, match="initial tau should be"):
+            cb = BYOLMAWeightUpdate(initial_tau)
+    else:
+        # Create simple one layer network and their copies
+        online_network = nn.Linear(100, 10)
+        target_network = deepcopy(online_network)
+        online_network_original = deepcopy(online_network)
+        target_network_original = deepcopy(target_network)
+
+        # Check parameters are equal
+        assert torch.equal(next(iter(online_network.parameters()))[0], next(iter(target_network.parameters()))[0])
+
+        # Simulate weight update
+        opt = torch.optim.SGD(online_network.parameters(), lr=0.1)
+        y = online_network(torch.randn(3, 100))
+        loss = y.sum()
+        loss.backward()
+        opt.step()
+        opt.zero_grad()
+
+        # Check online network update
+        assert not torch.equal(
+            next(iter(online_network.parameters()))[0], next(iter(online_network_original.parameters()))[0]
+        )
+
+        # Update target network weights via callback
+        cb = BYOLMAWeightUpdate(initial_tau)
+        cb.update_weights(online_network, target_network)
+
+        # Check target network update
+        if initial_tau == 0.0:
```

Comment:
can we check for other coefficients as well?

Reply:
I am not sure if I am understanding your suggestion correctly. The test currently checks the following cases:
- <0 or >1 raises warning
- equal to 0
- equal to 1
- everything else (0 < tau < 1)

Are there other cases to be checked?

Reply:
I mean, we know the values `target_network` weights should have, based on tau, since we did literally only one step. Therefore, we can check for the exact value based on tau, not only for the cases of `tau == 0` or `tau == 1`